website/docs/noreturn.md

```diff
@@ -3,40 +3,152 @@ id: noreturn
 
-sig {returns(T.noreturn)}
-def raise_always
-  raise RuntimeError
+Infinite loops like this occur frequently in long-lived processes that are
+designed to respond to user input indefinitely.
+
+Having type system support for this may not seem immediately useful at first,
+but ties into the previous section: Sorbet can detect code written after an
+infinite loop, and flag that the code is unreachable.
+
+## Detecting impossible-to-call methods
+
+Because of [intersection types](intersection-types.md), it is possible to create
+types for which there cannot be any values. If there are no values of a given
+type, and that type is used for an argument of a method, Sorbet flags the usage
+as dead code:
+
+```ruby
+class A; end
+class B; end
+
+sig {params(x: T.all(A, B)).void}
+def example(x)
+  puts(x) # error: This code is unreachable
 end
+```
+
+As discussed in
+[Understanding how intersection types collapse](intersection-types.md#understanding-how-intersection-types-collapse),
+`T.all(A, B)` collapses to `T.noreturn` because the classes `A` and `B` are
+completely unrelated, and it therefore impossible for a value to simultaneously
+be an instance of `A` and `B`.
+
+In these examples, Sorbet has caught an example of a method which is not
+possible to call.
+
+## `T.noreturn` and subtyping
+
+`T.noreturn` is a subtype of all other types. This can be useful in certain
+situations. For example, the type of the empty array is most accurately typed as
+`T::Array[T.noreturn]`. Because `T.noreturn` is a subtype of all other types,
+this means that `[]` has type `T::Array[Integer]` and `T::Array[String]` and
+`T::Array[T.any(Integer, String)]`, etc.
 
-x = exit_program
-puts x # error: This code is unreachable
+This also allows Sorbet to detect when, for example, there's an incorrect
+attempt to access the first element of a known-empty list:
+
+```ruby
+xs = T::Array[T.noreturn].new
+x = xs.fetch(0)
+puts(x) # error: This code is unreachable
+```
+
+## `T.noreturn` and `T.untyped`
+
+Despite what we've claimed throughout this page, technically it **is** possible
+to have a value of type `T.noreturn` by way of `T.untyped`:
+
+```ruby
+sig {returns(T.noreturn)}
+def always_raises
+  T.unsafe("Just kidding, it actually does return")
+end
 ```
 
-`T.noreturn` is a subtype of every type, but no type except `T.noreturn`
-(trivially) and `T.untyped` is a subtype of `T.noreturn`.
+This is an unavoidable downside of Sorbet's use of [`T.untyped`](untyped.md) to
+implement a [gradual type system](gradual.md).
+
+This is a major reason why Sorbet's [runtime type checking](runtime.md) is
+valuable: in order for Sorbet's dead code analysis to be accurate, Sorbet relies
+on the fact that violations of the declared types will as a last resort be
+caught at runtime by checking [method signatures](sigs.md) and
+[type assertions](type-assertions.md).
+
+Disabling these checks at runtime can cause Sorbet to incorrectly judge which
+code is or is not reachable at runtime.
```

Comment:
My understanding of this paragraph is that it's trying to avoid saying "Sorbet's (static) checks are not perfect, given the existence of `T.untyped`", but it winds up in a weird spot: the first part talks about disabling something at runtime, and the second part claims that this runtime disabling affects something happening at static typechecking time, which is backwards-in-time causality.

WDYT about rephrasing this?  I don't have great ideas after thinking about it for a couple of minutes.

Reply:
I took a pass at changing this.